Add support for custom ring key

asymmetric opened the pull request:
This PR adds support for specifying a custom ring key, generated by the user and then inserted into a Kubernetes secret.

Closes #41.

LiliC commented:
IIUC this is about securing the ring? But since our supervisors can't connect in a ring right now, as the work for that is still ongoing, can we even test this works?

asymmetric commented:
Good point. What we can see is that the the supervisor picks up the ring key as inteded: 

```bash
❯ kubectl logs example-encrypted-service-group-3058708116-dt3b4
∵ Missing package for core/hab-launcher        
» Installing core/hab-launcher                 
↓ Downloading core/hab-launcher/4571/20170727193541                                            
☛ Verifying core/hab-launcher/4571/20170727193541                                              
↓ Downloading core-20160810182414 public origin key                                            
☑ Cached core-20160810182414 public origin key 
→ Using core/bzip2/1.0.6/20170513212938        
→ Using core/cacerts/2017.01.18/20170513215104 
→ Using core/gcc-libs/5.2.0/20170513212920     
→ Using core/glibc/2.22/20170513201042         
→ Using core/libarchive/3.2.0/20170514021649   
→ Using core/libsodium/1.0.11/20170514010744   
→ Using core/linux-headers/4.3/20170513200956  
→ Using core/openssl/1.0.2j/20170513215106     
→ Using core/xz/5.2.2/20170513214327           
→ Using core/zlib/1.2.8/20170513201911         
✓ Installed core/hab-launcher/4571/20170727193541                                              
★ Install of core/hab-launcher/4571/20170727193541 complete with 1 new packages installed.     
hab-sup(MR): Joining ring example-encrypted-service-group-ring-key-20170822144706              
hab-sup(MR): Supervisor Member-ID ece548891cec49369d2f7836b70d7135                             
hab-sup(MR): Starting core/nginx               
hab-sup(MR): Starting gossip-listener on 0.0.0.0:9638                                          
hab-sup(MR): Starting http-gateway on 0.0.0.0:9631                                             
default(CF): Updated fastcgi.conf 588f1fc81fc357c200571835d3783cb9502ec61d52027c53be3fcc47aca9af31                                                                                             
default(CF): Updated fastcgi_params 5bbcc22e06b600a8b910a35303bdfa8f560f237e3d0108e6c5d0526914518cbf                                                                                           
default(CF): Updated win-utf a1956fd2fb8a8baadc78a0d7df22488b860bc1d8818253a1e54e499a5c9cda38  
default(CF): Updated koi-win bc10205f5978664b3e82ab760986fa939c08b2f11613a31a99b96a3ab1b83056  
default(CF): Updated uwsgi_params 7ccf985c8a6448a71f81920a76ef625e9f718c0ef563ed7674846ad80999bf1b                                                                                             
default(CF): Updated mime.types 4254386d16e9499ac06151beafa105b2123536d7481cb4d03b17ceae6a5dd568                                                                                               
default(CF): Updated nginx.conf 61a0ef13de8dc7fd0fee82df54920520688abbf95a159386cf7c4880e3564ccb                                                                                               
default(CF): Updated koi-utf 921e0547e5fb5c8ab85cdade1202d5b3bd21d616e54d483cd969ee03d7e9794f  
default(CF): Updated scgi_params bb76b03ae89d01ab33876311d0b3c8ec5e566e171ac258c4ed129aae66500c51                                                                                              
nginx.default(SR): Configuration recompiled    
nginx.default(SR): Initializing                
nginx.default(SV): Starting service as user=root, group=hab
```

LiliC commented:
Should we wait until that PR is merged, and merge this one then? WDYT? 

I don't know, on one hand, we know it works outside of Kubernetes, on the other hand we haven't tested with starting a full ring in our operator? Guess we might have similar problems in the future with being a bit ahead of Habitat work, so best to decide what to do here.

asymmetric commented:
Could I test this using the branch we use for demos?

LiliC commented:
Also does this always just secures a ring, or also individual supervisors? From the documentation I understand it just secures ring communication between supervisors. Do you know more? 

I am asking, because do we need this if topology is standalone?

asymmetric commented:
The above output is of a supervisor with no `-t` flag specified, so a standalone one.

So I think the answer to your question is that yes, this flag can also be used if the topology is standalone.

LiliC commented on pkg/habitat/controller/controller.go:
```diff
@@ -503,6 +508,55 @@ func (hc *HabitatController) newDeployment(sg *crv1.ServiceGroup) (*appsv1beta1.
+
+		// The current time. We need this because ring keys must have a revision.
+		// This means that the timestamp in the filename and the one in the ring
+		// key itself will be different, because the one in the key reflects the
```
We could get around this by just letting the user create a secret with name that is the timestamp, and we use that as our `ringKeyFile `. WDYT?

asymmetric replied:
Yes, we could do that, or we could add a `revision` key to the secret. I'm not sure what's best.

LiliC commented on pkg/habitat/controller/controller.go:
```diff
@@ -503,6 +508,55 @@ func (hc *HabitatController) newDeployment(sg *crv1.ServiceGroup) (*appsv1beta1.
+		vm := &apiv1.VolumeMount{
+			Name:      ringName,
+			MountPath: "/hab/cache/keys",
+			// This directory cannot be made read-only, as it is written to by the supervisor.
```
Not sure this comment is clear enough.

asymmetric commented on pkg/habitat/controller/controller.go:
```diff
@@ -42,6 +42,11 @@ const (
 	userTomlFile = "user.toml"
+
+	// The key under which the ring key is stored in the Kubernetes Secret.
+	ringSecretKey = "ring-key"
```
`ringKeyFileExt` is `sym.key` though. Did you mean some other constant?

asymmetric commented on examples/encrypted/README.md:
```diff
@@ -0,0 +1,21 @@
+
+## Deletion
+
+The operator does not delete the Secret on ServiceGroup deletion. This is
```
I think that might make sense in the main README, but not here, since we don't create any Services in this example and I'd find it confusing.

asymmetric commented:
Testing using krnowak's hab, it looks good:

```
/code/go/src/github.com/kinvolk/habitat-operator asymmetric/ring-key*
❯ kubectl logs example-encrypted-service-group-3232687993-g548g -f
hab-sup(MR): Joining ring example-encrypted-service-group-ring-key-20170823161628
hab-sup(MR): Supervisor Member-ID 382edba0ea954b69abc3ee130cbc703e
hab-sup(MR): Starting core/consul
hab-sup(MR): Starting gossip-listener on 0.0.0.0:9638
hab-sup(MR): Starting http-gateway on 0.0.0.0:9631
consul.default(SR): Hooks recompiled
default(CF): Updated basic_config.json 341b8c5aa6328ed464e6cc9419edd5805732ceb20818b70f94b7e398676e760a
consul.default(SR): Configuration recompiled
consul.default(SR): Waiting to execute hooks; election in progress, and we have no quorum.
consul.default(SR): Executing hooks; 7ea03a4a8f7b4e6ea4d9fc74d87b6eb5 is the leader
consul.default(SR): Initializing
consul.default(SV): Starting service as user=root, group=hab
```

I also addressed the comments, PTALA.

asymmetric commented:
Consul itself does not manage to elect a leader though, but that was the case in `lili/topology` as well IIRC.

asymmetric commented:
@LiliC I went with your suggestion of using the ring key as the secret's name. PTALA.

LiliC commented on examples/encrypted/README.md:
```diff
@@ -0,0 +1,29 @@
+# Encrypted ServiceGroup example
```
Btw 👍  for docs in our examples!

LiliC commented on pkg/habitat/controller/controller.go:
```diff
@@ -503,6 +508,48 @@ func (hc *HabitatController) newDeployment(sg *crv1.ServiceGroup) (*appsv1beta1.
+					SecretName: s.Name,
+					Items: []apiv1.KeyToPath{
+						{
+							Key:  ringSecretKey,
```
This is what I meant earlier, we can just reuse `ringKeyFile` here, as it `Key`. WDYT? Just to be consistent with what we already have.

asymmetric replied:
I'm not sure about that. This is a constant that references what key we want to access on the Secret, so it doesn't really deal with files.
